build-system/runner/src/org/ampproject/AmpPass.java

```diff
@@ -0,0 +1,103 @@
+package org.ampproject;
+
+import java.util.Set;
+
+import com.google.javascript.jscomp.AbstractCompiler;
+import com.google.javascript.jscomp.HotSwapCompilerPass;
+import com.google.javascript.jscomp.NodeTraversal;
+import com.google.javascript.jscomp.NodeTraversal.AbstractPostOrderCallback;
+import com.google.javascript.rhino.IR;
+import com.google.javascript.rhino.Node;
+
+/**
+ * Does a `stripTypeSuffix` which currently can't be done through
+ * the normal `strip` mechanisms provided by closure compiler.
+ * Some of the known mechanisms we tried before writing our own compiler pass
+ * are setStripTypes, setStripTypePrefixes, setStripNameSuffixes, setStripNamePrefixes.
+ * The normal mechanisms found in closure compiler can't strip the expressions we want because
+ * they are either prefix based and/or operate on the es6 translated code which would mean they
+ * operate on a qualifier string name that looks like
+ * "module$__$__$__$extensions$amp_test$0_1$log.dev.fine".
+ * 
+ * Other custom pass examples found inside closure compiler src:
+ * https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/PolymerPass.java
+ * https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/AngularPass.java
+ */
+class AmpPass extends AbstractPostOrderCallback implements HotSwapCompilerPass {
+
+  final AbstractCompiler compiler;
+  private final Set<String> stripTypeSuffixes;
+
+  public AmpPass(AbstractCompiler compiler, Set<String> stripTypeSuffixes) {
+    this.compiler = compiler;
+    this.stripTypeSuffixes = stripTypeSuffixes;
+  }
+
+  @Override public void process(Node externs, Node root) {
+    hotSwapScript(root, null);
+  }
+
+  @Override public void hotSwapScript(Node scriptRoot, Node originalRoot) {
+    NodeTraversal.traverseEs6(compiler, scriptRoot, this);
+  }
+
+  @Override public void visit(NodeTraversal t, Node n, Node parent) {
+    // If the line is a bare method call like `prop.method()` then
+    // the suffix name passed in is 'prop.method' but if there is an
+    // assignment like `var a = prop.method()` then the suffix name
+    // passed in will be 'a' which is how we are able to preserve
+    // var value = dev.assert(truthy, "message");
+    if (n.isExprResult()) {
+      maybeEliminateExpressionBySuffixName(t, n, parent);
+    }
+  }
+
+  /**
+   * Checks if expression is a GETPROP() (method invocation) and the property
+   * name ends with one of the items in stripTypeSuffixes.
+   */
+  private void maybeEliminateExpressionBySuffixName(NodeTraversal t, Node n, Node parent) {
+    // n = EXPRESSION_RESULT > CALL > GETPROP
+    Node call = n.getFirstChild();
+    if (call == null) {
+      return;
+    }
+    Node expression = call.getFirstChild();
+    if (expression == null) {
+      return;
+    }
+
+    if (qualifiedNameEndsWithStripType(expression)) {
+      if (parent.isExprResult()) {
```

Comment:
You currently just kill the full statement, right?

I wonder if, for asserts, you should just alway copy the first argument here. It would be DCEed if it is side effect free.